runtime/gc_vlhgc/IncrementalGenerationalGC.cpp

```diff
@@ -977,6 +977,9 @@ MM_IncrementalGenerationalGC::runGlobalMarkPhaseIncrement(MM_EnvironmentVLHGC *e
 	if ((_globalMarkPhaseIncrementBytesStillToScan > 0) || (MM_CycleState::state_process_work_packets_after_initial_mark != _persistentGlobalMarkPhaseState._markDelegateState)) {
 		globalMarkPhase(env, true);
 	}
+	else {
+		static_cast<MM_CycleStateVLHGC*>(env->_cycleState)->_vlhgcIncrementStats.clear();
```

Comment:
@amicic @LinHu2016 Could you please comment on wether clearing the increment stats here is correct. 

The stats do get cleared in line 978 `globalMarkPhase(env, true);`, but GMP will skip this if 

`if ((_globalMarkPhaseIncrementBytesStillToScan > 0) || (MM_CycleState::state_process_work_packets_after_initial_mark != _persistentGlobalMarkPhaseState._markDelegateState))` ends up being false

I think we need to clear the stats because it's possible to finish and report end of increment without ever clearing the increment stats for GMP, hence we can end up with old/stale stats of previous increment. This wasn't an issue before, but now that we use increment stats for reporting stall times, the old increment stall will be reported in the case `globalMarkPhase(env, true);` isn't called + I don't see any useful work being doing for which stats are collected when `globalMarkPhase(env, true)` is skipped. 

@gza060625 has verified that stats aren't cleared in this case.

Reply:
This looks correct to me. The only thing is, should we pull out that line in front of if statement (and remove it from globalMarkPhase)?

Reply:
This can be done, however `globalMarkPhase` method is used in one other place, for Global GC in `runGlobalGarbageCollection`.

We'll have to account for this clearing in runGlobalGarbageCollection as well then.

Reply:
`setupBeforeGlobalGC` is a common method called before `globalMarkPhase` for both GlobalGC and GMP, and its used exclusively by the 2. Perhaps the clearing can be moved up to there

Reply:
Only concern here would be is if anything between `setupBeforeGlobalGC` &  `globalMarkPhase` relies on increment stats being not cleared (e.g using previous increments stats for current increment etc), I don't see a reason for this.

```c++
MM_IncrementalGenerationalGC::runGlobalGarbageCollection(...) {
...
      setupBeforeGlobalGC(env, env->_cycleState->_gcCode); // **** MOVE CLEAR IN HERE ****

	if (_extensions->tarokUseProjectedSurvivalCollectionSet) {
		_projectedSurvivalCollectionSetDelegate.createRegionCollectionSetForGlobalGC(env);
	} else {
		_collectionSetDelegate.createRegionCollectionSetForGlobalGC(env);
	}
	
	_interRegionRememberedSet->prepareRegionsForGlobalCollect(env, isGlobalMarkPhaseRunning());

	globalMarkPhase(env, false); // **** EXISTING CLEAR ****
...
```


```c++
MM_IncrementalGenerationalGC::runGlobalMarkPhaseIncrement
...
	setupBeforeGlobalGC(env, env->_cycleState->_gcCode); // **** MOVE CLEAR IN HERE ****

	if(!isGlobalMarkPhaseRunning()) {
		reportGMPCycleStart(env);
		_persistentGlobalMarkPhaseState._vlhgcCycleStats.clear();
	}

	reportGMPIncrementStart(env);
	reportGCIncrementStart(env, "GMP increment", env->_cycleState->_currentIncrement);

	_extensions->globalVLHGCStats.gcCount += 1;

	if ((_globalMarkPhaseIncrementBytesStillToScan > 0) || (MM_CycleState::state_process_work_packets_after_initial_mark != _persistentGlobalMarkPhaseState._markDelegateState)) {
		globalMarkPhase(env, true); // **** EXISTING CLEAR ****
	}
...
```

Reply:
I'm not super familiar with the code, would like to confirm this change or simply take path of least changes. 

Another "safer" solution is pull out the clear from` globalMarkPhase` and add explicitly have the clear before calling `globalMarkPhase` at both call sites.

Reply:
Those are good observations. Moving it to setupBeforeGlobalGC sounds very reasonable (simple and safe).

BTW, there is another point where clearing is done in IIGC::internalPreCollect. I believe  this path is taken when doing final STW increment of GMP. It may become redundant if we make introduce a clear for any increment of GMP. Hard to say without actually running and verifying it.

Reply:
@gza060625 this PR can be wrapped up this with the changes discussed above